workspace/src/main/kotlin/com/cosmotech/workspace/azure/WorkspaceServiceImpl.kt

```diff
@@ -43,94 +37,101 @@ import org.springframework.stereotype.Service
+    }
+  }
 
   override fun findAllWorkspaces(organizationId: String) =
       cosmosTemplate.findAll<Workspace>("${organizationId}_workspaces")
 
-  override fun findWorkspaceById(organizationId: String, workspaceId: String): Workspace =
-      cosmosTemplate.findByIdOrThrow(
-          "${organizationId}_workspaces",
-          workspaceId,
-          "Workspace $workspaceId not found in organization $organizationId")
+  override fun findWorkspaceById(organizationId: String, workspaceId: String): Workspace {
+    val workspace: Workspace =
+        cosmosTemplate.findByIdOrThrow(
+            "${organizationId}_workspaces",
+            workspaceId,
+            "Workspace $workspaceId not found in organization $organizationId")
+    this.validateUser(workspace)
+    return workspace
+  }
 
   override fun removeAllUsersOfWorkspace(organizationId: String, workspaceId: String) {
     val workspace = findWorkspaceById(organizationId, workspaceId)
+    if (workspace.ownerId != getCurrentAuthenticatedUserName()) {
+      // TODO Only the owner or an admin should be able to perform this operation
+      throw CsmAccessForbiddenException("You are not allowed to update this Resource")
+    }
     if (!workspace.users.isNullOrEmpty()) {
-      val userIds = workspace.users!!.mapNotNull { it.id }
       workspace.users = listOf()
       cosmosTemplate.upsert("${organizationId}_workspaces", workspace)
-
-      userIds.forEach {
-        this.eventPublisher.publishEvent(
-            UserRemovedFromWorkspace(this, organizationId, workspaceId, it))
-      }
     }
   }
 
-  override fun removeUserFromOrganizationWorkspace(
+  override fun removeUserFromWorkspace(
       organizationId: String,
       workspaceId: String,
-      userId: String
+      userMail: String
   ) {
     val workspace = findWorkspaceById(organizationId, workspaceId)
-    val workspaceUserMap =
-        workspace.users?.associateBy { it.id!! }?.toMutableMap() ?: mutableMapOf()
-    if (workspaceUserMap.containsKey(userId)) {
-      workspaceUserMap.remove(userId)
-      workspace.users = workspaceUserMap.values.toList()
-      cosmosTemplate.upsert("${organizationId}_workspaces", workspace)
-      this.eventPublisher.publishEvent(
-          UserRemovedFromWorkspace(this, organizationId, workspaceId, userId))
+    if (workspace.ownerId != getCurrentAuthenticatedUserName()) {
+      // TODO Only the owner or an admin should be able to perform this operation
+      throw CsmAccessForbiddenException("You are not allowed to update this Resource")
     }
-  }
+    workspace.users = workspace.users?.filter { it != userMail }
 
-  override fun addOrReplaceUsersInOrganizationWorkspace(
-      organizationId: String,
-      workspaceId: String,
-      workspaceUser: List<WorkspaceUser>
-  ): List<WorkspaceUser> {
-    if (workspaceUser.isEmpty()) {
-      // Nothing to do
-      return workspaceUser
-    }
+    cosmosTemplate.upsert("${organizationId}_workspaces", workspace)
```

Comment:
Great, but we could optimize this by performing this potentially expensive `upsert` operation only if the updated list of users has really changed.

Reply:
todo

Reply:
done